blog/content/post/2021-02-05-testing.md

```diff
@@ -0,0 +1,134 @@
+amount of time, instead of clumsy and dangerous release. In this blogpost we
+will try to narrow down the distinction to the core, as comparing manual and
+automated testing. The former is the simpler method and does not add the cost of
+writing code, so may be performed immediately. The later is based on long term
+thinking about the problem, where scripts perform the tests without human
+participation. And let's be clear – such investment into writing comes with many
+further benefits.
+
+
+#### 3mdeb validation approach
+
+How are manual tests different from automated tests? Manual testing is based on
+performing the role of a standard user – no code is written. For obvious
+reasons, the whole process takes much longer and, although initially it does not
+require large investments, results in much lower ROI, due to the costs related
+to human resources. In some cases, manual testing is an option worth
+considering, for example when performing usability, or in no-scenario tests.
+However, in reproducible tasks, like functional tests, only the automation will
+bring the expected results.
+
+![testing_chart](/img/test1.png)
+
+Our validation infrastructure in 3mdeb consists of ~150 tests. Performing the
+whole testing process step by step would take days, especially that the whole
+process is repetitive. Summarizing, manual testing in this case costs more and
+slows down the whole development pace, when great delay occurs between code
+release and feedback for the programmers. If we would still try to test
+manually, our knowledge of code quality would be worse and the delivery would be
+delayed.
+
+The whole process may be performed in a much more convenient way. Tests can be
+run by the developer, who needs only a brief information about the testing
+environment. Clear documentation and ready-to-use physical platforms with system
+of automatic availability management make it possible to run both, the single
+tests and complete suites, concerning a set of functionalities that must be
+checked. In result we receive immediate feedback about the changes made and the
+features that were added.
+
+#### Remote Testing Environment
+
+As we are performing our test on hardware we had to found a way to automate
+management and operations on physical platforms. To meet that goal we have
+created Remote Testing Environment - a Orange Pi Zero hat that gives us
+ability to remotely control power, simulate pressing of buttons like power on
+or reset, read serial output and flash firmware using physical cable connection
+that can unbrick device without any manual intervention. This setup has a
+corresponding configuration file for every platform model which allow for
+smooth remote test execution. You can find [RTE](https://jira-3mdeb.atlassian.net/browse/SHA-195?focusedCommentId=55760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-55760) in our shop.
```

Comment:
@KBanecka link to be replaced for correct one

Reply:
done